platforms/reference/src/SimTKReference/ReferenceVelocityVerletDynamics.cpp

```diff
@@ -0,0 +1,274 @@
+#include "ReferenceVirtualSites.h"
+
+#include <cstdio>
+
+using std::vector;
+using namespace OpenMM;
+
+/**---------------------------------------------------------------------------------------
+
+   ReferenceVelocityVerletDynamics constructor
+
+   @param numberOfAtoms  number of atoms
+   @param deltaT         delta t for dynamics
+   @param friction       friction coefficient
+   @param temperature    temperature
+
+   --------------------------------------------------------------------------------------- */
+
+ReferenceVelocityVerletDynamics::ReferenceVelocityVerletDynamics(int numberOfAtoms, double deltaT) :
+           ReferenceDynamics(numberOfAtoms, deltaT, 0.0) {
+   xPrime.resize(numberOfAtoms);
+   inverseMasses.resize(numberOfAtoms);
+}
+
+/**---------------------------------------------------------------------------------------
+
+   ReferenceVelocityVerletDynamics destructor
+
+   --------------------------------------------------------------------------------------- */
+
+ReferenceVelocityVerletDynamics::~ReferenceVelocityVerletDynamics() {
+}
+
+/**---------------------------------------------------------------------------------------
+
+   Update -- driver routine for performing Velocity Verlet dynamics update of coordinates
+   and velocities
+
+   @param system              the System to be integrated
+   @param atomCoordinates     atom coordinates
+   @param velocities          velocities
+   @param forces              forces
+   @param masses              atom masses
+   @param atomList            list of all atoms not involved in a Drude-like pair
+   @param pairList            list of all Drude-like pairs
+   @param maxPairDistance     the maximum separation of any Drude-like pairs
+
+   --------------------------------------------------------------------------------------- */
+
+void ReferenceVelocityVerletDynamics::update(OpenMM::ContextImpl &context, const OpenMM::System& system, vector<Vec3>& atomCoordinates,
+                                          vector<Vec3>& velocities,
+                                          vector<Vec3>& forces, vector<double>& masses, double tolerance, bool &forcesAreValid,
+                                          const std::vector<int> & atomList, const std::vector<std::tuple<int, int, double>> &pairList,
+                                          double maxPairDistance) {
+
+    // first-time-through initialization
+    if (!forcesAreValid) context.calcForcesAndEnergy(true, false);
+
+    int numberOfAtoms = system.getNumParticles();
+    if (getTimeStep() == 0) {
+       // invert masses
+
+       for (int ii = 0; ii < numberOfAtoms; ii++) {
+          if (masses[ii] == 0.0)
+              inverseMasses[ii] = 0.0;
+          else
+              inverseMasses[ii] = 1.0/masses[ii];
+       }
+    }
+
+    //// Perform the integration.
+
+    // Regular atoms
+    for (const auto &atom : atomList) {
+        if (masses[atom] != 0.0) {
+            for (int xyz = 0; xyz < 3; ++xyz) {
```

Comment:
You don't actually need all these xyz loops.  Vec3 supports standard math operations.  For example, you could just write

```c++

velocities[atom] += 0.5*inverseMasses[atom]*forces[atom]*getDeltaT();
xPrime[atom] = atomCoordinates[atom];
atomCoordinates[atom] += velocities[atom]*getDeltaT();
```